File: tools/conformance/versions.mjs
import assert from 'node:assert/strict';
import { mkdtemp } from 'node:fs/promises';
import { join } from 'node:path';
import { Database } from '../../packages/node/dist/index.js';
import { sdk, daemon, cli, mcp } from './adapters.mjs';
import { command, bounded } from './process.mjs';

const high = 9007199254740993n;
const maximum = 9223372036854775807n;
const u64Max = 18446744073709551615n;
const batch = (id, version) => ({ op: 'batch', operations: [
  { model: 'kv_set', namespace: 'versions', key: [...Buffer.from(id)], value: [43], if_version: version.toString(), expires_at_ms: 4102444800000 },
  { model: 'json_put', space: 'versions', id, value: { number: 4294967297 }, if_version: version.toString() },
] });

async function verify(surface, execute) {
  const call = async step => {
    try { return await bounded(execute(step), `${surface} precision ${step.op}`); }
    catch (error) {
      if (error instanceof Error) error.message = `${surface} ${step.op}: ${error.message}`;
      throw error;
    }
  };
  const getKv = id => call({ op: 'kv_get', namespace: 'versions', key: [...Buffer.from(id)] });
  const getJson = id => call({ op: 'json_get', space: 'versions', id });
  const rejected = (step, code, details) => assert.rejects(call(step), error => {
    assert.equal(error.code, code, `${surface}: error code`);
    assert.equal(error.retryable, code === 'CONFLICT');
    assert.equal(typeof (error.requestId ?? error.request_id), 'string');
    if (details) assert.deepEqual(error.details, details, `${surface}: exact conflict details`);
    return true;
  });
  assert.equal((await getKv('high')).version, high.toString());
  const document = await getJson('high');
  assert.equal(document.version, high.toString());
  assert.deepEqual(document.value, { number: 4294967297 });
  assert.equal((await getKv('high')).expires_at_ms, 4102444800000);
  assert.deepEqual(await call(batch('high', high)), [
    { version: (high + 1n).toString(), deleted: false }, { version: (high + 1n).toString(), deleted: false },
  ]);
  for (const expected of [high, u64Max]) {
    await rejected(batch('high', expected), 'CONFLICT', { expected: expected.toString(), actual: (high + 1n).toString() });
  }
  const rollback = batch('high', high);
  rollback.operations.unshift({ model: 'json_put', space: 'versions', id: 'rollback', value: true });
  await rejected(rollback, 'CONFLICT');
  assert.equal(await getJson('rollback'), null);
  assert.equal((await getKv('high')).version, (high + 1n).toString());
  for (const op of ['kv_scan', 'json_scan']) {
    const page = await call({ op, namespace: 'versions', space: 'versions', after: op === 'kv_scan' ? [...Buffer.from('edge')] : 'edge', limit: 1 });
    assert.equal(page[0].version, (high + 1n).toString());
  }
  assert.deepEqual(await call(batch('edge', maximum - 1n)), [
    { version: maximum.toString(), deleted: false }, { version: maximum.toString(), deleted: false },
  ]);
  await rejected(batch('edge', maximum), 'STORAGE');
  assert.equal((await getJson('edge')).version, maximum.toString());
  // JSON/native parsing and SDK range checks must reject, never wrap or round.
  for (const version of [-1n, u64Max + 1n]) await rejected(batch('high', version), 'INVALID_INPUT');
  assert.equal((await getJson('high')).version, (high + 1n).toString());
  assert.deepEqual(await call({ op: 'batch', operations: [
    { model: 'kv_delete', namespace: 'versions', key: [...Buffer.from('high')], if_version: (high + 1n).toString() },
    { model: 'json_delete', space: 'versions', id: 'high', if_version: (high + 1n).toString() },
  ] }), [{ version: null, deleted: true }, { version: null, deleted: true }]);
  assert.equal(await getKv('high'), null);
  assert.equal(await getJson('high'), null);
  console.log(`${surface}: exact versions >2^53, u64::MAX conflicts, i64::MAX exhaustion passed`);
}

export async function precision(parent, executable) {
  const directory = await mkdtemp(join(parent, 'precision-'));
  const result = command('cargo', ['test', '-p', 'lilia-storage-sqlite', '--test', 'versions', '--', '--include-ignored'], {
    stdio: 'inherit', timeout: 0, env: { ...process.env, LILIA_VERSION_FIXTURE_DIR: directory },
  });
  assert.equal(result.status, 0, 'seed version fixtures and validate Rust embedded');
  const embedded = await Database.open({ path: join(directory, 'embedded.lilia') });
  try { await verify('Node embedded', sdk(embedded)); }
  finally { await embedded.close(); }
  const shared = await daemon(executable('lilia-daemon'), directory);
  try { await verify('daemon', shared.execute); }
  finally { await shared.close(); }
  await verify('CLI JSON', cli(executable('lilia'), join(directory, 'cli.lilia'), directory));
  const server = await mcp(executable('lilia-mcp'), join(directory, 'mcp.lilia'));
  try { await verify('MCP stdio', server.execute); }
  finally { await server.close(); }
}

File: crates/lilia-storage-sqlite/tests/versions.rs
use lilia_core::{BatchOperation, ErrorCode};
use lilia_storage_sqlite::{Database, DatabaseOptions};
use rusqlite::{params, Connection};
use std::path::Path;
use std::time::Duration;

const HIGH: u64 = (1_u64 << 53) + 1;

fn put(id: &str, version: Option<u64>) -> [BatchOperation; 2] {
    [
        BatchOperation::KvSet {
            namespace: "versions".into(),
            key: id.as_bytes().to_vec(),
            value: vec![42],
            if_version: version,
            expires_at_ms: Some(4_102_444_800_000),
        },
        BatchOperation::JsonPut {
            space: "versions".into(),
            id: id.into(),
            value: serde_json::json!({"number": 4_294_967_297_u64}),
            if_version: version,
        },
    ]
}

fn seed(path: &Path) {
    assert!(
        !path.exists(),
        "fixture must never overwrite an existing database"
    );
    let database = Database::open(DatabaseOptions::durable(path)).unwrap();
    for id in ["high", "edge"] {
        database.batch(&put(id, None)).unwrap();
    }
    database.close(Duration::from_secs(5)).unwrap();
    let connection = Connection::open(path).unwrap();
    for (id, version) in [
        ("high", i64::try_from(HIGH).unwrap()),
        ("edge", i64::MAX - 1),
    ] {
        connection
            .execute(
                "UPDATE _lilia_kv SET version=?1 WHERE namespace='versions' AND key=?2",
                params![version, id.as_bytes()],
            )
            .unwrap();
        connection
            .execute(
                "UPDATE _lilia_json SET version=?1 WHERE space='versions' AND id=?2",
                params![version, id],
            )
            .unwrap();
    }
    connection.close().unwrap();
}

#[test]
#[ignore = "private fixture producer invoked by the five-surface conformance runner"]
fn seed_precision_fixture() {
    let directory = std::env::var_os("LILIA_VERSION_FIXTURE_DIR").expect("fixture directory");
    for name in ["embedded", "daemon", "cli", "mcp"] {
        seed(&Path::new(&directory).join(format!("{name}.lilia")));
    }
}

#[test]
fn high_versions_conflicts_and_exhaustion_are_exact() {
    let directory = tempfile::tempdir().unwrap();
    let path = directory.path().join("db.lilia");
    seed(&path);
    let database = Database::open(DatabaseOptions::durable(&path)).unwrap();
    assert_eq!(
        database
            .kv_get("versions", b"high")
            .unwrap()
            .unwrap()
            .version,
        HIGH
    );
    assert_eq!(
        database
            .json_get("versions", "high")
            .unwrap()
            .unwrap()
            .version,
        HIGH
    );
    assert_eq!(
        database.batch(&put("high", Some(HIGH))).unwrap()[0].version,
        Some(HIGH + 1)
    );
    let error = database.batch(&put("high", Some(u64::MAX))).unwrap_err();
    assert_eq!(error.code, ErrorCode::Conflict);
    assert_eq!(
        error.details.unwrap(),
        serde_json::json!({"expected": u64::MAX.to_string(), "actual": (HIGH + 1).to_string()})
    );
    let maximum = u64::try_from(i64::MAX).unwrap();
    database.batch(&put("edge", Some(maximum - 1))).unwrap();
    let mut operations = put("rolled-back", None).to_vec();
    operations.extend(put("edge", Some(maximum)));
    let error = database.batch(&operations).unwrap_err();
    assert_eq!(error.code, ErrorCode::Storage);
    assert!(!error.retryable);
    assert!(database
        .json_get("versions", "rolled-back")
        .unwrap()
        .is_none());
    assert!(database
        .kv_get("versions", b"rolled-back")
        .unwrap()
        .is_none());
    database.close(Duration::from_secs(5)).unwrap();
    let database = Database::open(DatabaseOptions::durable(&path)).unwrap();
    assert_eq!(
        database.kv_scan("versions", Some(b"edge"), 1).unwrap()[0].version,
        HIGH + 1
    );
    assert_eq!(
        database.json_scan("versions", Some("edge"), 1).unwrap()[0].version,
        HIGH + 1
    );
    assert!(database.integrity_check().unwrap());
}
